subsys/settings/src/settings_shell.c

```diff
@@ -114,11 +114,46 @@ static int cmd_read(const struct shell *shell_ptr, size_t argc, char *argv[])
 	return 0;
 }
 
+static int cmd_write(const struct shell *shell_ptr, size_t argc, char *argv[])
+{
+	int err;
+	uint8_t buffer[CONFIG_SHELL_CMD_BUFF_SIZE / 2];
+	size_t buffer_len;
+
+	buffer_len = hex2bin(argv[2], strlen(argv[2]), buffer, sizeof(buffer));
+
+	if (buffer_len == 0) {
+		err = -ENOMEM;
+	} else {
+		err = settings_save_one(argv[1], buffer, buffer_len);
+	}
```

Comment:
As far as I understand  https://github.com/zephyrproject-rtos/zephyr/pull/45459#issuecomment-1124600652 the `buffer_len` should never return 0 because the length of buffers will always be OK; the only reason to get 0 here is then when there is even number of hex digits (for example 0AB can not be converted to bin). This means that `-ENOMEM` may not be proper error code here.
Additionally the `settings_save_one` will be calling backend that may return `-ENOMEM`, whit means that at line 132 we do not know which function actually failed.
Shouldn't we ensure that `hex2bin` gets fed proper hex string so that we could issue error that parameter is not OK and just check result of `settings_save_one` for the "Failed to write ..." error message?

Reply:
Ok, I can return `-EINVAL` in such a case, too. Regarding ambiguity of error codes argument, isn't this a "problem" of majority of the Zephyr (and not only Zephyr) codebase? In many cases where you get an error code, you don't know which internal component returned it. I don't quite understand why it started being a problem in my PR :).

Reply:
> Ok, I can return `-EINVAL` in such a case, too. Regarding ambiguity of error codes argument, isn't this a "problem" of majority of the Zephyr (and not only Zephyr) codebase? In many cases where you get an error code, you don't know which internal component returned it. I don't quite understand why it started being a problem in my PR :).

What I mean is that by knowing the issue with `hex2bin`, and assuming that you will always give it a correct buffer size, the only reason to get 0 is uneven hex string, so maybe just do:
```
if (buffer_len == 0) {
   // Report problem with user input
   shell_error("Uneven hex value");
   err = -EINVAL;
} else {
   err =  ..
}
shell_error("Failed to ... ");
```
or something like that?

Reply:
Done.